Use nested schemas

I noticed that we're using Terraform's map type to represent Kafka brokers:

https://github.com/MaterializeInc/terraform-provider-materialize/blob/e20b557841486c9ea2dfc6c845796669555db46f/materialize/resources/resource_connection.go#L169

but then reaching inside the map to pull out expected fields:

https://github.com/MaterializeInc/terraform-provider-materialize/blob/e20b557841486c9ea2dfc6c845796669555db46f/materialize/resources/resource_connection.go#L564-L588

This is going to fail with an ugly panic if a user typos a map field in their HCL.

A better approach is to use nested schemas inside of a list. See the `ordered_cache_behavior` field in the `aws_cloudfront_distribution` resource for an example: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudfront_distribution

```
resource "aws_cloudfront_distribution" "s3_distribution" {
  # Cache behavior with precedence 0
  ordered_cache_behavior {
    path_pattern     = "/content/immutable/*"
    allowed_methods  = ["GET", "HEAD", "OPTIONS"]
    cached_methods   = ["GET", "HEAD", "OPTIONS"]
    target_origin_id = local.s3_origin_id

    forwarded_values {
      query_string = false
      headers      = ["Origin"]

      cookies {
        forward = "none"
      }
    }

    min_ttl                = 0
    default_ttl            = 86400
    max_ttl                = 31536000
    compress               = true
    viewer_protocol_policy = "redirect-to-https"
  }

  # Cache behavior with precedence 1
  ordered_cache_behavior {
    path_pattern     = "/content/*"
    allowed_methods  = ["GET", "HEAD", "OPTIONS"]
    cached_methods   = ["GET", "HEAD"]
    target_origin_id = local.s3_origin_id

    forwarded_values {
      query_string = false

      cookies {
        forward = "none"
      }
    }

    min_ttl                = 0
    default_ttl            = 3600
    max_ttl                = 86400
    compress               = true
    viewer_protocol_policy = "redirect-to-https"
  }
}
```

This allows Terraform's type checking to kick in, and filter out invalid fields, etc. 

Here's a similar example from the Frontegg provider that shows how this is implemented: https://github.com/frontegg/terraform-provider-frontegg/blob/9976fb90bb60aaa28fed945ee67aefb761bd8e23/provider/resource_frontegg_workspace.go#L544-L566

I've been stuck on this for a bit now, as soon as I change ` Elem:          &schema.Schema{Type: schema.TypeMap},` to a nested schema, I start hitting this error:

```
2023-02-23T17:54:22.975+0200 [ERROR] vertex "materialize_connection.example_kafka_connection" error: Unsupported argument
╷
│ Error: Unsupported argument
│ 
│   on main.tf line 87, in resource "materialize_connection" "example_kafka_connection":
│   87:   kafka_brokers = [{
│ 
│ An argument named "kafka_brokers" is not expected here. 
```

The same argument is accepted just fine when using `TypeMap`.

Draft PR:

https://github.com/MaterializeInc/terraform-provider-materialize/pull/39

--- a/examples/main.tf
+++ b/examples/main.tf
@@ -79,16 +79,16 @@ resource "materialize_connection" "privatelink_conn" {
   aws_privatelink_service_name       = "com.amazonaws.us-east-1.materialize.example"
   aws_privatelink_availability_zones = ["use1-az2", "use1-az1"]
 }
-resource "materialize_connection" "example_kafka_connection" {
-  name            = "example_kafka_connection"
+resource "materialize_connection" "example_kafka_privatelink_conn" {
+  name            = "example_kafka_privatelink_conn"
   connection_type = "KAFKA"
-  kafka_brokers {
+  kafka_broker {
     broker                 = "b-1.hostname-1:9096"
     target_group_port      = "9001"
     availability_zone      = "use1-az1"
     privatelink_connection = "privatelink_conn"
   }
-  kafka_brokers {
+  kafka_broker {
     broker                 = "b-2.hostname-2:9096"
     target_group_port      = "9002"
     availability_zone      = "use1-az2"
@@ -112,10 +112,10 @@ resource "materialize_connection" "example_postgres_connection" {
 resource "materialize_connection" "example_kafka_connection" {
   name            = "example_kafka_connection"
   connection_type = "KAFKA"
-  kafka_brokers {
+  kafka_broker {
     broker = "b-1.hostname-1:9096"
   }
-  kafka_brokers {
+  kafka_broker {
     broker = "b-2.hostname-2:9096"
   }
   kafka_sasl_username   = "example"

--- a/examples/main.tf
+++ b/examples/main.tf
@@ -72,27 +72,30 @@ resource "materialize_connection" "example_ssh_connection" {
 # # Create a AWS Private Connection
 # Note: you need the max_aws_privatelink_connections increased for this to work:
 # show max_aws_privatelink_connections;
-# resource "materialize_connection" "example_kafka_connection" {
-#   name            = "example_kafka_connection"
-#   connection_type = "KAFKA"
-#   # kafka_broker    = "example.com:9092"
-#   kafka_brokers = [{
-#     "broker" : "b-1.hostname-1:9096",
-#     "target_group_port" : "9001",
-#     "availability_zone" : "use1-az1",
-#     "privatelink_connection" : "privatelink_conn",
-#     },
-#     {
-#       "broker" : "b-2.hostname-2:9096",
-#       "target_group_port" : "9002",
-#       "availability_zone" : "use1-az2",
-#       "privatelink_connection" : "privatelink_conn",
-#   }]
-#   kafka_sasl_username   = "example"
-#   kafka_sasl_password   = "kafka_password"
-#   kafka_sasl_mechanisms = "SCRAM-SHA-256"
-#   kafka_progress_topic  = "example"
-# }
+resource "materialize_connection" "privatelink_conn" {
+  name                               = "privatelink_conn"
+  schema_name                        = "public"
+  connection_type                    = "AWS PRIVATELINK"
+  aws_privatelink_service_name       = "com.amazonaws.us-east-1.materialize.example"
+  aws_privatelink_availability_zones = ["use1-az2", "use1-az1"]
+}
+resource "materialize_connection" "example_kafka_connection" {
+  name            = "example_kafka_connection"
+  connection_type = "KAFKA"
+  kafka_brokers {
+    broker = "b-1.hostname-1:9096"
+    target_group_port = "9001"
+    availability_zone = "use1-az1"
+    privatelink_connection = "privatelink_conn"
+  }
+  kafka_brokers {
+      broker = "b-2.hostname-2:9096"
+      target_group_port = "9002"
+      availability_zone = "use1-az2"
+      privatelink_connection = "privatelink_conn"
+  }
+  depends_on = [materialize_connection.privatelink_conn]
+}
 
 # Create a Postgres Connection
 resource "materialize_connection" "example_postgres_connection" {
@@ -109,7 +112,6 @@ resource "materialize_connection" "example_postgres_connection" {
 resource "materialize_connection" "example_kafka_connection" {
   name            = "example_kafka_connection"
   connection_type = "KAFKA"
-  # kafka_broker    = "example.com:9092"
   kafka_brokers = [{
     "broker" : "b-1.hostname-1:9096",
     },

--- a/examples/main.tf
+++ b/examples/main.tf
@@ -83,16 +83,16 @@ resource "materialize_connection" "example_kafka_connection" {
   name            = "example_kafka_connection"
   connection_type = "KAFKA"
   kafka_brokers {
-    broker = "b-1.hostname-1:9096"
-    target_group_port = "9001"
-    availability_zone = "use1-az1"
+    broker                 = "b-1.hostname-1:9096"
+    target_group_port      = "9001"
+    availability_zone      = "use1-az1"
     privatelink_connection = "privatelink_conn"
   }
   kafka_brokers {
-      broker = "b-2.hostname-2:9096"
-      target_group_port = "9002"
-      availability_zone = "use1-az2"
-      privatelink_connection = "privatelink_conn"
+    broker                 = "b-2.hostname-2:9096"
+    target_group_port      = "9002"
+    availability_zone      = "use1-az2"
+    privatelink_connection = "privatelink_conn"
   }
   depends_on = [materialize_connection.privatelink_conn]
 }
@@ -112,12 +112,12 @@ resource "materialize_connection" "example_postgres_connection" {
 resource "materialize_connection" "example_kafka_connection" {
   name            = "example_kafka_connection"
   connection_type = "KAFKA"
-  kafka_brokers = [{
-    "broker" : "b-1.hostname-1:9096",
-    },
-    {
-      "broker" : "b-2.hostname-2:9096",
-  }]
+  kafka_brokers {
+    broker = "b-1.hostname-1:9096"
+  }
+  kafka_brokers {
+    broker = "b-2.hostname-2:9096"
+  }
   kafka_sasl_username   = "example"
   kafka_sasl_password   = "kafka_password"
   kafka_sasl_mechanisms = "SCRAM-SHA-256"

--- a/examples/resources/materialize_connection/resource.tf
+++ b/examples/resources/materialize_connection/resource.tf
@@ -54,7 +54,7 @@ resource "materialize_connection" "example_postgres_connection" {
 resource "materialize_connection" "example_kafka_connection" {
   name            = "example_kafka_connection"
   connection_type = "KAFKA"
-  kafka_brokers {
+  kafka_broker {
     broker = "b-1.hostname-1:9096"
   }
   kafka_sasl_username   = "example"
@@ -89,13 +89,13 @@ resource "materialize_connection" "example_confluent_schema_registry_connection"
 resource "materialize_connection" "example_kafka_connection" {
   name            = "example_kafka_connection"
   connection_type = "KAFKA"
-  kafka_brokers {
+  kafka_broker {
     broker = "b-1.hostname-1:9096"
     target_group_port = "9001"
     availability_zone = "use1-az1"
     privatelink_connection = "privatelink_conn"
   }
-  kafka_brokers {
+  kafka_broker {
     broker = "b-2.hostname-2:9096"
     target_group_port = "9002"
     availability_zone = "use1-az2"

--- a/examples/resources/materialize_connection/resource.tf
+++ b/examples/resources/materialize_connection/resource.tf
@@ -49,17 +49,14 @@ resource "materialize_connection" "example_postgres_connection" {
 #     DATABASE 'example'
 # );
 
+
 # Create a Kafka Connection
 resource "materialize_connection" "example_kafka_connection" {
   name            = "example_kafka_connection"
   connection_type = "KAFKA"
-  kafka_broker    = "example.com:9092"
-  # kafka_brokers         = [{
-  #   "broker": "b-1.hostname-1:9096",
-  # },
-  # {
-  #   "broker": "b-2.hostname-2:9096",
-  # }]
+  kafka_brokers {
+    broker = "b-1.hostname-1:9096"
+  }
   kafka_sasl_username   = "example"
   kafka_sasl_password   = "kafka_password"
   kafka_sasl_mechanisms = "SCRAM-SHA-256"
@@ -92,19 +89,18 @@ resource "materialize_connection" "example_confluent_schema_registry_connection"
 resource "materialize_connection" "example_kafka_connection" {
   name            = "example_kafka_connection"
   connection_type = "KAFKA"
-  # kafka_broker    = "example.com:9092"
-  kafka_brokers = [{
-    "broker" : "b-1.hostname-1:9096",
-    "target_group_port" : "9001",
-    "availability_zone" : "use1-az1",
-    "privatelink_connection" : "privatelink_conn",
-    },
-    {
-      "broker" : "b-2.hostname-2:9096",
-      "target_group_port" : "9002",
-      "availability_zone" : "use1-az2",
-      "privatelink_connection" : "privatelink_conn",
-  }]
+  kafka_brokers {
+    broker = "b-1.hostname-1:9096"
+    target_group_port = "9001"
+    availability_zone = "use1-az1"
+    privatelink_connection = "privatelink_conn"
+  }
+  kafka_brokers {
+    broker = "b-2.hostname-2:9096"
+    target_group_port = "9002"
+    availability_zone = "use1-az2"
+    privatelink_connection = "privatelink_conn"
+  }
 }
 
 # CREATE CONNECTION materialize.public.example_kafka_connection TO KAFKA (

--- a/integration/connection.tf
+++ b/integration/connection.tf
@@ -9,7 +9,9 @@ resource "materialize_connection" "ssh_connection" {
 resource "materialize_connection" "kafka_connection" {
   name            = "kafka_connection"
   connection_type = "KAFKA"
-  kafka_broker    = "kafka:9092"
+  kafka_brokers {
+    broker = "kafka:9092"
+  }
 }
 
 resource "materialize_connection" "schema_registry" {

--- a/integration/connection.tf
+++ b/integration/connection.tf
@@ -9,7 +9,7 @@ resource "materialize_connection" "ssh_connection" {
 resource "materialize_connection" "kafka_connection" {
   name            = "kafka_connection"
   connection_type = "KAFKA"
-  kafka_brokers {
+  kafka_broker {
     broker = "kafka:9092"
   }
 }
@@ -32,10 +32,10 @@ resource "materialize_connection" "example_ssh_connection" {
 resource "materialize_connection" "kafka_conn_multiple_brokers" {
   name            = "kafka_conn_multiple_brokers"
   connection_type = "KAFKA"
-  kafka_brokers {
+  kafka_broker {
     broker = "kafka:9092"
   }
-  kafka_brokers {
+  kafka_broker {
     broker = "kafka2:9092"
   }
   kafka_sasl_username   = "sasl_user"

--- a/materialize/resources/resource_connection.go
+++ b/materialize/resources/resource_connection.go
@@ -297,6 +297,13 @@ func Connection() *schema.Resource {
 	}
 }
 
+type KafkaBroker struct {
+	Broker                string
+	TargetGroupPort       int
+	AvailabilityZone      string
+	PrivateLinkConnection string
+}
+
 type ConnectionBuilder struct {
 	connectionName                        string
 	schemaName                            string
@@ -318,7 +325,7 @@ type ConnectionBuilder struct {
 	postgresSSLKey                        string
 	postgresSSLMode                       string
 	postgresAWSPrivateLink                string
-	kafkaBrokers                          []map[string]interface{}
+	kafkaBrokers                          []KafkaBroker
 	kafkaProgressTopic                    string
 	kafkaSSLCa                            string
 	kafkaSSLCert                          string
@@ -440,7 +447,7 @@ func (b *ConnectionBuilder) PostgresAWSPrivateLink(postgresAWSPrivateLink string
 	return b
 }
 
-func (b *ConnectionBuilder) KafkaBrokers(kafkaBrokers []map[string]interface{}) *ConnectionBuilder {
+func (b *ConnectionBuilder) KafkaBrokers(kafkaBrokers []KafkaBroker) *ConnectionBuilder {
 	b.kafkaBrokers = kafkaBrokers
 	return b
 }
@@ -574,7 +581,7 @@ func (b *ConnectionBuilder) Create() string {
 			if b.kafkaSSHTunnel != "" {
 				q.WriteString(`BROKERS (`)
 				for i, broker := range b.kafkaBrokers {
-					q.WriteString(fmt.Sprintf(`'%s' USING SSH TUNNEL %s`, broker["broker"], b.kafkaSSHTunnel))
+					q.WriteString(fmt.Sprintf(`'%s' USING SSH TUNNEL %s`, broker.Broker, b.kafkaSSHTunnel))
 					if i < len(b.kafkaBrokers)-1 {
 						q.WriteString(`,`)
 					}
@@ -583,13 +590,13 @@ func (b *ConnectionBuilder) Create() string {
 			} else {
 				q.WriteString(fmt.Sprintf(`BROKERS (`))
 				for i, broker := range b.kafkaBrokers {
-					if broker["target_group_port"] != nil && broker["target_group_port"] != 0 && broker["availability_zone"] != nil && broker["availability_zone"] != "" && broker["privatelink_connection"] != nil && broker["privatelink_connection"] != "" {
-						q.WriteString(fmt.Sprintf(`'%s' USING AWS PRIVATELINK %s (PORT %d, AVAILABILITY ZONE '%s')`, broker["broker"], broker["privatelink_connection"], broker["target_group_port"], broker["availability_zone"]))
+					if broker.TargetGroupPort != 0 && broker.AvailabilityZone != "" && broker.PrivateLinkConnection != "" {
+						q.WriteString(fmt.Sprintf(`'%s' USING AWS PRIVATELINK %s (PORT %d, AVAILABILITY ZONE '%s')`, broker.Broker, broker.PrivateLinkConnection, broker.TargetGroupPort, broker.AvailabilityZone))
 						if i < len(b.kafkaBrokers)-1 {
 							q.WriteString(`, `)
 						}
 					} else {
-						q.WriteString(fmt.Sprintf(`'%s'`, broker["broker"]))
+						q.WriteString(fmt.Sprintf(`'%s'`, broker.Broker))
 						if i < len(b.kafkaBrokers)-1 {
 							q.WriteString(`, `)
 						}
@@ -789,9 +796,15 @@ func connectionCreate(ctx context.Context, d *schema.ResourceData, meta interfac
 	}
 
 	if v, ok := d.GetOk("kafka_broker"); ok {
-		brokers := []map[string]interface{}{}
-		for _, b := range v.([]interface{}) {
-			brokers = append(brokers, b.(map[string]interface{}))
+		var brokers []KafkaBroker
+		for _, broker := range v.([]interface{}) {
+			b := broker.(map[string]interface{})
+			brokers = append(brokers, KafkaBroker{
+				Broker:                b["broker"].(string),
+				TargetGroupPort:       b["target_group_port"].(int),
+				AvailabilityZone:      b["availability_zone"].(string),
+				PrivateLinkConnection: b["privatelink_connection"].(string),
+			})
 		}
 		builder.KafkaBrokers(brokers)
 	}

--- a/materialize/resources/resource_connection.go
+++ b/materialize/resources/resource_connection.go
@@ -162,7 +162,7 @@ var connectionSchema = map[string]*schema.Schema{
 		Type:        schema.TypeString,
 		Optional:    true,
 	},
-	"kafka_brokers": {
+	"kafka_broker": {
 		Description: "The Kafka brokers configuration.",
 		Type:        schema.TypeList,
 		Optional:    true,
@@ -787,7 +787,7 @@ func connectionCreate(ctx context.Context, d *schema.ResourceData, meta interfac
 		builder.PostgresSSHTunnel(v.(string))
 	}
 
-	if v, ok := d.GetOk("kafka_brokers"); ok {
+	if v, ok := d.GetOk("kafka_broker"); ok {
 		brokers := []map[string]interface{}{}
 		for _, b := range v.([]interface{}) {
 			brokers = append(brokers, b.(map[string]interface{}))

--- a/materialize/resources/resource_connection.go
+++ b/materialize/resources/resource_connection.go
@@ -166,6 +166,7 @@ var connectionSchema = map[string]*schema.Schema{
 		Description: "The Kafka brokers configuration.",
 		Type:        schema.TypeList,
 		Optional:    true,
+		MinItems:    1,
 		Elem: &schema.Resource{
 			Schema: map[string]*schema.Schema{
 				"broker": {

--- a/materialize/resources/resource_connection.go
+++ b/materialize/resources/resource_connection.go
@@ -582,7 +582,7 @@ func (b *ConnectionBuilder) Create() string {
 			} else {
 				q.WriteString(fmt.Sprintf(`BROKERS (`))
 				for i, broker := range b.kafkaBrokers {
-					if broker["target_group_port"] != nil && broker["availability_zone"] != nil && broker["privatelink_connection"] != nil {
+					if broker["target_group_port"] != nil && broker["target_group_port"] != 0 && broker["availability_zone"] != nil && broker["availability_zone"] != "" && broker["privatelink_connection"] != nil && broker["privatelink_connection"] != "" {
 						q.WriteString(fmt.Sprintf(`'%s' USING AWS PRIVATELINK %s (PORT %d, AVAILABILITY ZONE '%s')`, broker["broker"], broker["privatelink_connection"], broker["target_group_port"], broker["availability_zone"]))
 						if i < len(b.kafkaBrokers)-1 {
 							q.WriteString(`, `)

--- a/materialize/resources/resource_connection.go
+++ b/materialize/resources/resource_connection.go
@@ -166,9 +166,32 @@ var connectionSchema = map[string]*schema.Schema{
 	"kafka_brokers": {
 		Description:   "The Kafka brokers configuration.",
 		Type:          schema.TypeList,
-		Elem:          &schema.Schema{Type: schema.TypeMap},
 		Optional:      true,
 		ConflictsWith: []string{"kafka_broker"},
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"broker": {
+					Description: "The Kafka broker, in the form of `host:port`.",
+					Type:        schema.TypeString,
+					Required:    true,
+				},
+				"target_group_port": {
+					Description: "The port of the target group associated with the Kafka broker.",
+					Type:        schema.TypeInt,
+					Optional:    true,
+				},
+				"availability_zone": {
+					Description: "The availability zone of the Kafka broker.",
+					Type:        schema.TypeString,
+					Optional:    true,
+				},
+				"privatelink_connection": {
+					Description: "The AWS PrivateLink connection name in Materialize.",
+					Type:        schema.TypeString,
+					Optional:    true,
+				},
+			},
+		},
 	},
 	"kafka_progress_topic": {
 		Description: "The name of a topic that Kafka sinks can use to track internal consistency metadata.",
@@ -786,7 +809,6 @@ func connectionCreate(ctx context.Context, d *schema.ResourceData, meta interfac
 		for _, b := range v.([]interface{}) {
 			brokers = append(brokers, b.(map[string]interface{}))
 		}
-		// panic(fmt.Sprintf("kafka_brokers: %v, %T", brokers, brokers))
 		builder.KafkaBrokers(brokers)
 	}
 

--- a/materialize/resources/resource_connection.go
+++ b/materialize/resources/resource_connection.go
@@ -157,17 +157,10 @@ var connectionSchema = map[string]*schema.Schema{
 		Type:        schema.TypeString,
 		Optional:    true,
 	},
-	"kafka_broker": {
-		Description:   "The Kafka broker configuration.",
-		Type:          schema.TypeString,
-		Optional:      true,
-		ConflictsWith: []string{"kafka_brokers"},
-	},
 	"kafka_brokers": {
-		Description:   "The Kafka brokers configuration.",
-		Type:          schema.TypeList,
-		Optional:      true,
-		ConflictsWith: []string{"kafka_broker"},
+		Description: "The Kafka brokers configuration.",
+		Type:        schema.TypeList,
+		Optional:    true,
 		Elem: &schema.Resource{
 			Schema: map[string]*schema.Schema{
 				"broker": {
@@ -319,7 +312,6 @@ type ConnectionBuilder struct {
 	postgresSSLKey                        string
 	postgresSSLMode                       string
 	postgresAWSPrivateLink                string
-	kafkaBroker                           string
 	kafkaBrokers                          []map[string]interface{}
 	kafkaProgressTopic                    string
 	kafkaSSLCa                            string
@@ -439,11 +431,6 @@ func (b *ConnectionBuilder) PostgresSSLMode(postgresSSLMode string) *ConnectionB
 
 func (b *ConnectionBuilder) PostgresAWSPrivateLink(postgresAWSPrivateLink string) *ConnectionBuilder {
 	b.postgresAWSPrivateLink = postgresAWSPrivateLink
-	return b
-}
-
-func (b *ConnectionBuilder) KafkaBroker(kafkaBroker string) *ConnectionBuilder {
-	b.kafkaBroker = kafkaBroker
 	return b
 }
 
@@ -577,12 +564,6 @@ func (b *ConnectionBuilder) Create() string {
 	}
 
 	if b.connectionType == "KAFKA" {
-		if b.kafkaBroker != "" {
-			q.WriteString(fmt.Sprintf(`BROKER '%s'`, b.kafkaBroker))
-			if b.kafkaSSHTunnel != "" {
-				q.WriteString(fmt.Sprintf(` USING SSH TUNNEL %s`, b.kafkaSSHTunnel))
-			}
-		}
 		if len(b.kafkaBrokers) != 0 {
 			if b.kafkaSSHTunnel != "" {
 				q.WriteString(`BROKERS (`)
@@ -597,7 +578,7 @@ func (b *ConnectionBuilder) Create() string {
 				q.WriteString(fmt.Sprintf(`BROKERS (`))
 				for i, broker := range b.kafkaBrokers {
 					if broker["target_group_port"] != nil && broker["availability_zone"] != nil && broker["privatelink_connection"] != nil {
-						q.WriteString(fmt.Sprintf(`'%s' USING AWS PRIVATELINK %s (PORT %s, AVAILABILITY ZONE '%s')`, broker["broker"], broker["privatelink_connection"], broker["target_group_port"], broker["availability_zone"]))
+						q.WriteString(fmt.Sprintf(`'%s' USING AWS PRIVATELINK %s (PORT %d, AVAILABILITY ZONE '%s')`, broker["broker"], broker["privatelink_connection"], broker["target_group_port"], broker["availability_zone"]))
 						if i < len(b.kafkaBrokers)-1 {
 							q.WriteString(`, `)
 						}
@@ -798,10 +779,6 @@ func connectionCreate(ctx context.Context, d *schema.ResourceData, meta interfac
 
 	if v, ok := d.GetOk("postgres_ssh_tunnel"); ok {
 		builder.PostgresSSHTunnel(v.(string))
-	}
-
-	if v, ok := d.GetOk("kafka_broker"); ok {
-		builder.KafkaBroker(v.(string))
 	}
 
 	if v, ok := d.GetOk("kafka_brokers"); ok {

--- a/materialize/resources/resource_connection_test.go
+++ b/materialize/resources/resource_connection_test.go
@@ -48,9 +48,9 @@ func TestResourceConnectionCreateKafka(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBrokers([]map[string]interface{}{
+	b.KafkaBrokers([]KafkaBroker{
 		{
-			"broker": "localhost:9092",
+			Broker: "localhost:9092",
 		},
 	})
 	b.KafkaProgressTopic("topic")
@@ -64,12 +64,12 @@ func TestResourceConnectionCreateKafkaMultipleBrokers(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBrokers([]map[string]interface{}{
+	b.KafkaBrokers([]KafkaBroker{
 		{
-			"broker": "localhost:9092",
+			Broker: "localhost:9092",
 		},
 		{
-			"broker": "localhost:9093",
+			Broker: "localhost:9093",
 		},
 	})
 	b.KafkaProgressTopic("topic")
@@ -83,9 +83,9 @@ func TestResourceConnectionCreateKafkaSsh(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBrokers([]map[string]interface{}{
+	b.KafkaBrokers([]KafkaBroker{
 		{
-			"broker": "localhost:9092",
+			Broker: "localhost:9092",
 		},
 	})
 	b.KafkaProgressTopic("topic")
@@ -100,12 +100,12 @@ func TestResourceConnectionCreateKafkaBrokers(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBrokers([]map[string]interface{}{
+	b.KafkaBrokers([]KafkaBroker{
 		{
-			"broker": "localhost:9092",
+			Broker: "localhost:9092",
 		},
 		{
-			"broker": "localhost:9093",
+			Broker: "localhost:9093",
 		},
 	})
 	b.KafkaProgressTopic("topic")
@@ -119,12 +119,12 @@ func TestResourceConnectionCreateKafkaBrokersSsh(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBrokers([]map[string]interface{}{
+	b.KafkaBrokers([]KafkaBroker{
 		{
-			"broker": "localhost:9092",
+			Broker: "localhost:9092",
 		},
 		{
-			"broker": "localhost:9093",
+			Broker: "localhost:9093",
 		},
 	})
 	b.KafkaProgressTopic("topic")
@@ -139,9 +139,9 @@ func TestResourceConnectionCreateKafkaSsl(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBrokers([]map[string]interface{}{
+	b.KafkaBrokers([]KafkaBroker{
 		{
-			"broker": "localhost:9092",
+			Broker: "localhost:9092",
 		},
 	})
 	b.KafkaProgressTopic("topic")
@@ -166,18 +166,18 @@ func TestResourceConnectionKafkaAwsPrivatelink(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBrokers([]map[string]interface{}{
+	b.KafkaBrokers([]KafkaBroker{
 		{
-			"broker":                 "b-1.hostname-1:9096",
-			"target_group_port":      9001,
-			"availability_zone":      "use1-az1",
-			"privatelink_connection": "privatelink_conn",
+			Broker:                "b-1.hostname-1:9096",
+			TargetGroupPort:       9001,
+			AvailabilityZone:      "use1-az1",
+			PrivateLinkConnection: "privatelink_conn",
 		},
 		{
-			"broker":                 "b-1.hostname-1:9097",
-			"target_group_port":      9002,
-			"availability_zone":      "use1-az2",
-			"privatelink_connection": "privatelink_conn",
+			Broker:                "b-1.hostname-1:9097",
+			TargetGroupPort:       9002,
+			AvailabilityZone:      "use1-az2",
+			PrivateLinkConnection: "privatelink_conn",
 		},
 	})
 	b.KafkaSASLMechanisms("PLAIN")

--- a/materialize/resources/resource_connection_test.go
+++ b/materialize/resources/resource_connection_test.go
@@ -48,12 +48,16 @@ func TestResourceConnectionCreateKafka(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBroker("localhost:9092")
+	b.KafkaBrokers([]map[string]interface{}{
+		{
+			"broker": "localhost:9092",
+		},
+	})
 	b.KafkaProgressTopic("topic")
 	b.KafkaSASLMechanisms("PLAIN")
 	b.KafkaSASLUsername("user")
 	b.KafkaSASLPassword("password")
-	r.Equal(`CREATE CONNECTION database.schema.kafka_conn TO KAFKA (BROKER 'localhost:9092', PROGRESS TOPIC 'topic', SASL MECHANISMS = 'PLAIN', SASL USERNAME = 'user', SASL PASSWORD = SECRET password);`, b.Create())
+	r.Equal(`CREATE CONNECTION database.schema.kafka_conn TO KAFKA (BROKERS ('localhost:9092'), PROGRESS TOPIC 'topic', SASL MECHANISMS = 'PLAIN', SASL USERNAME = 'user', SASL PASSWORD = SECRET password);`, b.Create())
 }
 
 func TestResourceConnectionCreateKafkaMultipleBrokers(t *testing.T) {
@@ -79,13 +83,17 @@ func TestResourceConnectionCreateKafkaSsh(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBroker("localhost:9092")
+	b.KafkaBrokers([]map[string]interface{}{
+		{
+			"broker": "localhost:9092",
+		},
+	})
 	b.KafkaProgressTopic("topic")
 	b.KafkaSASLMechanisms("PLAIN")
 	b.KafkaSASLUsername("user")
 	b.KafkaSASLPassword("password")
 	b.KafkaSSHTunnel("ssh_conn")
-	r.Equal(`CREATE CONNECTION database.schema.kafka_conn TO KAFKA (BROKER 'localhost:9092' USING SSH TUNNEL ssh_conn, PROGRESS TOPIC 'topic', SASL MECHANISMS = 'PLAIN', SASL USERNAME = 'user', SASL PASSWORD = SECRET password);`, b.Create())
+	r.Equal(`CREATE CONNECTION database.schema.kafka_conn TO KAFKA (BROKERS ('localhost:9092' USING SSH TUNNEL ssh_conn), PROGRESS TOPIC 'topic', SASL MECHANISMS = 'PLAIN', SASL USERNAME = 'user', SASL PASSWORD = SECRET password);`, b.Create())
 }
 
 func TestResourceConnectionCreateKafkaBrokers(t *testing.T) {
@@ -131,12 +139,16 @@ func TestResourceConnectionCreateKafkaSsl(t *testing.T) {
 	r := require.New(t)
 	b := newConnectionBuilder("kafka_conn", "schema", "database")
 	b.ConnectionType("KAFKA")
-	b.KafkaBroker("localhost:9092")
+	b.KafkaBrokers([]map[string]interface{}{
+		{
+			"broker": "localhost:9092",
+		},
+	})
 	b.KafkaProgressTopic("topic")
 	b.KafkaSSLKey("key")
 	b.KafkaSSLCert("cert")
 	b.KafkaSSLCa("ca")
-	r.Equal(`CREATE CONNECTION database.schema.kafka_conn TO KAFKA (BROKER 'localhost:9092', PROGRESS TOPIC 'topic', SSL CERTIFICATE AUTHORITY = SECRET ca, SSL CERTIFICATE = SECRET cert, SSL KEY = SECRET key);`, b.Create())
+	r.Equal(`CREATE CONNECTION database.schema.kafka_conn TO KAFKA (BROKERS ('localhost:9092'), PROGRESS TOPIC 'topic', SSL CERTIFICATE AUTHORITY = SECRET ca, SSL CERTIFICATE = SECRET cert, SSL KEY = SECRET key);`, b.Create())
 }
 
 func TestResourceConnectionCreateConfluentSchemaRegistry(t *testing.T) {
@@ -157,13 +169,13 @@ func TestResourceConnectionKafkaAwsPrivatelink(t *testing.T) {
 	b.KafkaBrokers([]map[string]interface{}{
 		{
 			"broker":                 "b-1.hostname-1:9096",
-			"target_group_port":      "9001",
+			"target_group_port":      9001,
 			"availability_zone":      "use1-az1",
 			"privatelink_connection": "privatelink_conn",
 		},
 		{
 			"broker":                 "b-1.hostname-1:9097",
-			"target_group_port":      "9002",
+			"target_group_port":      9002,
 			"availability_zone":      "use1-az2",
 			"privatelink_connection": "privatelink_conn",
 		},

